test/integration_test.py

```diff
@@ -2054,6 +2055,13 @@ def test_response_security_headers(self):
             '/oauth2_redirect': {'Cache-Control': 'no-store'},
             '/health/basic': {'Cache-Control': 'no-store'}
         }
+        # https://www.w3.org/TR/CSP2/#policy-syntax
+        directive_re = re.compile(r'[ \t]*([a-zA-Z0-9-]+)'
+                                  # Directive value can be any visible character
+                                  # (0x21-0xFE), plus ' ' (0x20) and '\t' (0x09)
+                                  # except ',' (0x2C) and ';' (0x3B).
+                                  r'(?:[ \t]([\x21-\x2b\x2d-\x3a\x3c-\xFE\x20\x09]+))?')
```

Comment:
1) You are mixing lower and upper case hex digits. 

2) The `?` anchor is redundant with `fullmatch()` and even if it weren't you would need to anchor at the beginning as well `^`. 

3) The grammar specifies that the value can be empty. You use `+`.

4) The comment is too verbose.

5) There is no need to specify space or tab in hex, especially considering that you then need to explain it in the comment and that the preceding separator pattern uses the shorter form.

So ultimately


```diff
directive_re = re.compile(r'^[ \t]*([a-zA-Z0-9-]+)'
                          # Space, tab and any visible character (0x21-0xFE) 
                          # except for comma (0x2C) or semicolon (0x3B).
                          r'(?:[ \t]([ \t\x21-\x2B\x2D-\x3A\x3C-\xFE]*))?')
```

Reply:
> The `?` anchor is redundant with `fullmatch()` and even if it weren't you would need to anchor at the beginning as well `^`.

I believe you are confusing `?` with `$`. The `?` is being used to make the non-capturing group optional.

Reply:
> The grammar specifies that the value can be empty. You use +.

`directive-token = *WSP [ directive-name [ WSP directive-value ] ]`

I'm using `+` on the `directive-value` regex inside the optional non-capturing group so that if a WSP occurs after the `directive-name` a directive-value is then required.

Reply:
> There is no need to specify space or tab in hex, especially considering that you then need to explain it in the comment and that the preceding separator pattern uses the shorter form.

This was a result of testing the regex in the console. Using a mix of hex & escape characters produced a warning:

```
>>> re.fullmatch('[ \t\x21-\x2B\x2D-\x3A\x3C-\xFE]', 'foo')
<input>:1: FutureWarning: Possible set difference at position 6
```

While using only hex did not:
```
>>> re.fullmatch('[\x20\x09\x21-\x2B\x2D-\x3A\x3C-\xFE]', 'foo')
```

This doesn't occur with compiled regular expressions however so I will use the mix.

```
test_re = re.compile('[ \t\x21-\x2B\x2D-\x3A\x3C-\xFE]')
test_re.fullmatch('foo')
```